File: java/oauth2/src/main/java/com/rbleggi/oauth2/Main.java
package com.rbleggi.oauth2;

import java.util.Map;

public class Main {
    public static void main(String[] args) {
        System.out.println("OAuth2");
    }
}

interface GrantHandler {
    Map<String, String> handle(Map<String, String> request);
}

class AuthorizationCodeHandler implements GrantHandler {
    @Override
    public Map<String, String> handle(Map<String, String> request) {
        var code = "auth_code_123";
        return Map.of("code", code);
    }
}

class ClientCredentialsHandler implements GrantHandler {
    @Override
    public Map<String, String> handle(Map<String, String> request) {
        var token = "access_token_abc";
        return Map.of("access_token", token, "token_type", "bearer");
    }
}

class GrantHandlerFactory {
    static GrantHandler getHandler(String grantType) {
        return switch (grantType) {
            case "authorization_code" -> new AuthorizationCodeHandler();
            case "client_credentials" -> new ClientCredentialsHandler();
            default -> null;
        };
    }
}

class OAuth2Server {
    static Map<String, String> processRequest(Map<String, String> params) {
        var grantType = params.getOrDefault("grant_type", "");
        var handler = GrantHandlerFactory.getHandler(grantType);
        return handler != null ? handler.handle(params) : Map.of("error", "unsupported_grant_type");
    }
}

File: kotlin/oauth2/src/main/kotlin/com/rbleggi/oauth2/Main.kt
package com.rbleggi.oauth2

interface GrantHandler {
    fun handle(request: Map<String, String>): Map<String, String>
}

class AuthorizationCodeHandler : GrantHandler {
    override fun handle(request: Map<String, String>): Map<String, String> {
        val code = "auth_code_123"
        return mapOf("code" to code)
    }
}

class ClientCredentialsHandler : GrantHandler {
    override fun handle(request: Map<String, String>): Map<String, String> {
        val token = "access_token_abc"
        return mapOf("access_token" to token, "token_type" to "bearer")
    }
}

object GrantHandlerFactory {
    fun getHandler(grantType: String): GrantHandler? = when (grantType) {
        "authorization_code" -> AuthorizationCodeHandler()
        "client_credentials" -> ClientCredentialsHandler()
        else -> null
    }
}

object OAuth2Server {
    fun processRequest(params: Map<String, String>): Map<String, String> {
        val grantType = params["grant_type"] ?: ""
        return GrantHandlerFactory.getHandler(grantType)
            ?.handle(params)
            ?: mapOf("error" to "unsupported_grant_type")
    }
}

fun main() {
    println("OAuth2")
}
